metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,278 @@
+import os
+import sys
+import time
+import signal
+import shutil
+import hashlib
+import asyncio
+import tempfile
+from typing import List
+
+
+def hash_command_invocation(command: List[str]):
+    concatenated_string = "".join(command)
+    current_time = str(time.time())
+    concatenated_string += current_time
+    hash_object = hashlib.sha256(concatenated_string.encode())
+    return hash_object.hexdigest()
+
+
+class LogReadTimeoutError(Exception):
+    pass
+
+
+class SubprocessManager(object):
+    def __init__(self):
+        self.commands = {}
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.cleanup()
+
+    async def run_command(self, command: List[str], env=None, cwd=None):
+        command_id = hash_command_invocation(command)
+        self.commands[command_id] = CommandManager(command, env, cwd)
+        await self.commands[command_id].run()
+        return command_id
+
+    def get(self, command_id: str) -> "CommandManager":
+        return self.commands.get(command_id, None)
+
+    async def cleanup(self):
+        for _, v in self.commands.items():
+            await v.cleanup()
+
+
+class CommandManager(object):
+    def __init__(self, command: List[str], env=None, cwd=None):
+        self.command = command
+
+        if env is None:
+            env = os.environ.copy()
+        self.env = env
+
+        if cwd is None:
+            cwd = os.getcwd()
+        self.cwd = cwd
+
+        self.process = None
+        self.run_called = False
+        self.log_files = {}
+
+        signal.signal(signal.SIGINT, self.handle_sigint)
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.cleanup()
+
+    def handle_sigint(self, signum, frame):
+        print("SIGINT received.")
+        asyncio.create_task(self.kill())
+
+    async def wait(self, timeout=None, stream=None):
+        if timeout is None:
+            if stream is None:
+                await self.process.wait()
+            else:
+                await self.emit_logs(stream)
+        else:
+            tasks = [asyncio.create_task(asyncio.sleep(timeout))]
```

Comment:
way of doing the same thing at line 143 seems cleaner :)

Reply:
thanks, done